### PR TITLE
[Validator] ensure the proper context for nested validations

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14639, #20368
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by HeahDude** [comment]
A necessary feature since 2.7 is LTS and this bug's really tricky
👍

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php [hunk 1]
```diff
@@ -110,6 +111,11 @@ public function validate($value, $constraints = null, $groups = null)
         $previousMetadata = $this->context->getMetadata();
         $previousPath = $this->context->getPropertyPath();
         $previousGroup = $this->context->getGroup();
+        $previousConstraint = null;
+
+        if ($this->context instanceof ExecutionContext || method_exists($this->context, 'getConstraint')) {
```
Comment (ogizanagi) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: Why not just `if (method_exists($this->context, 'getConstraint'))` ?

Comment (xabbuh) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: I did not test it, but the `instanceof` check should be faster and should be true most of the time.

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**